RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h

```diff
@@ -91,13 +112,56 @@ class GlobalTrackingRegionWithVerticesProducer : public TrackingRegionProducer
-	  result.push_back( std::make_unique<GlobalTrackingRegion>(thePtMin, theOrigin_, theOriginRadius, theOriginHalfLength_, thePrecise, theUseMS) );
-	  if(theMaxNVertices >= 0 && result.size() >= static_cast<unsigned>(theMaxNVertices))
-	    break;
+
+          //scaling origin radius, half length, min pt for high-occupancy HI events to keep timing reasonable
+          if(theOriginRScaling || thePtMinScaling || theHalfLengthScaling){
+            //Use the unscaled radius unless one of the two conditions below is met
+            double scaledOriginRadius = theOriginRadius;
+            double scaledHalfLength   = theFixedError;
+            double scaledPtMin        = thePtMin;
+
+            //calculate nPixels (adapted from TkSeedGenerator/src/ClusterChecker.cc)
+            double nPix = 0;
+            edm::Handle<edmNew::DetSetVector<SiPixelCluster> > pixelClusterDSV;
+            ev.getByToken(token_pc, pixelClusterDSV);
+            if (!pixelClusterDSV.failedToGet()) {
+              const edmNew::DetSetVector<SiPixelCluster> & input = *pixelClusterDSV;
+              nPix = input.dataSize();
+            }
+            else{
+              edm::LogError("GlobalTrackingRegionProducerFromVertex")<<"could not get any SiPixel cluster collections of type edm::DetSetVector<SiPixelCluster>";
+              nPix = theScalingEnd+1;//if can't find collection, default to minimum radius to be safe
+            } 
+           
+ 
+            //first condition is for high occupancy, second makes sure we won't divide by zero or a negative number
+            if((nPix > theScalingEnd) || ((theScalingEnd-theScalingStart) <= 0)){
+              if(theOriginRScaling)    scaledOriginRadius = theMinOriginR;   // sets parameters to minimum value from PSet
+              if(theHalfLengthScaling) scaledHalfLength = theMinHalfLength;
+              if(thePtMinScaling)      scaledPtMin = theMaxPtMin; 
+            }
+            //second condition - scale radius linearly by Npix in the region from ScalingStart to ScalingEnd
+            else if((nPix <= theScalingEnd) && (nPix > theScalingStart)){
+              if(theOriginRScaling) scaledOriginRadius = theOriginRadius - (theOriginRadius-theMinOriginR)*(nPix-theScalingStart)/(theScalingEnd-theScalingStart);
+              if(theHalfLengthScaling) scaledHalfLength = theFixedError - (theFixedError-theMinHalfLength)*(nPix-theScalingStart)/(theScalingEnd-theScalingStart);
+              if(thePtMinScaling) scaledPtMin = thePtMin - (thePtMin-theMaxPtMin)*(nPix-theScalingStart)/(theScalingEnd-theScalingStart);
+            }
+            //std::cout << "NumberOfPixels: " <<  nPix << std::endl;
+            //std::cout << "Scaled Origin R: " << scaledOriginRadius << " Default Origin R: " << theOriginRadius <<" Scaled HalfLength: " << scaledHalfLength << " Default H.L.: " << theFixedError <<" Scaled pT Min: " << scaledPtMin << " Default pT Min: " << thePtMin << std::endl;
```

Comment:
please cleanup the commented out debugs.
If they are still needed, better enable them with a config parameter or use LogDebug